File: sdk/javascript/src/utils/encoding.ts
/**
 * Encoding utilities for DeshChain SDK
 */

import { toBase64, fromBase64, toHex, fromHex } from '@cosmjs/encoding'

export class EncodingUtils {
  /**
   * Convert string to base64
   */
  static toBase64(data: string): string {
    return toBase64(new TextEncoder().encode(data))
  }

  /**
   * Convert base64 to string
   */
  static fromBase64(data: string): string {
    return new TextDecoder().decode(fromBase64(data))
  }

  /**
   * Convert bytes to hex string
   */
  static toHex(data: Uint8Array): string {
    return toHex(data)
  }

  /**
   * Convert hex string to bytes
   */
  static fromHex(data: string): Uint8Array {
    return fromHex(data)
  }

  /**
   * Convert string to bytes
   */
  static stringToBytes(data: string): Uint8Array {
    return new TextEncoder().encode(data)
  }

  /**
   * Convert bytes to string
   */
  static bytesToString(data: Uint8Array): string {
    return new TextDecoder().decode(data)
  }

  /**
   * Encode JSON to base64
   */
  static jsonToBase64(data: any): string {
    return EncodingUtils.toBase64(JSON.stringify(data))
  }

  /**
   * Decode JSON from base64
   */
  static base64ToJson<T = any>(data: string): T {
    return JSON.parse(EncodingUtils.fromBase64(data))
  }

  /**
   * Create message hash
   */
  static hashMessage(message: string): string {
    // Simple hash implementation
    let hash = 0
    for (let i = 0; i < message.length; i++) {
      const char = message.charCodeAt(i)
      hash = ((hash << 5) - hash) + char
      hash = hash & hash // Convert to 32-bit integer
    }
    return Math.abs(hash).toString(16)
  }

  /**
   * Validate address format
   */
  static isValidAddress(address: string, prefix: string = 'deshchain'): boolean {
    if (!address.startsWith(prefix)) return false
    if (address.length !== prefix.length + 39) return false
    
    const addressPart = address.slice(prefix.length)
    return /^[a-z0-9]+$/.test(addressPart)
  }

  /**
   * Validate transaction hash
   */
  static isValidTxHash(hash: string): boolean {
    return /^[A-F0-9]{64}$/i.test(hash)
  }

  /**
   * Format amount with decimals
   */
  static formatAmount(amount: number | string, decimals: number = 6): string {
    const num = typeof amount === 'string' ? parseFloat(amount) : amount
    return (num / Math.pow(10, decimals)).toFixed(decimals)
  }

  /**
   * Parse amount to micro units
   */
  static parseAmount(amount: number | string, decimals: number = 6): string {
    const num = typeof amount === 'string' ? parseFloat(amount) : amount
    return Math.floor(num * Math.pow(10, decimals)).toString()
  }

  /**
   * Normalize denom
   */
  static normalizeDenom(denom: string): string {
    return denom.toLowerCase()
  }

  /**
   * Create random ID
   */
  static generateId(length: number = 16): string {
    const chars = 'abcdefghijklmnopqrstuvwxyz0123456789'
    let result = ''
    for (let i = 0; i < length; i++) {
      result += chars.charAt(Math.floor(Math.random() * chars.length))
    }
    return result
  }
}

File: sdk/python/deshchain/utils/encoding.py
"""
Encoding Utilities

Helper functions for encoding/decoding data.
"""

import base64
import json
import hashlib
from typing import Any, Dict


class EncodingUtils:
    """Encoding utility functions."""
    
    @staticmethod
    def to_base64(data: str) -> str:
        """Convert string to base64."""
        return base64.b64encode(data.encode()).decode()
    
    @staticmethod
    def from_base64(data: str) -> str:
        """Convert base64 to string."""
        return base64.b64decode(data).decode()
    
    @staticmethod
    def to_hex(data: bytes) -> str:
        """Convert bytes to hex string."""
        return data.hex()
    
    @staticmethod
    def from_hex(data: str) -> bytes:
        """Convert hex string to bytes."""
        return bytes.fromhex(data)
    
    @staticmethod
    def json_to_base64(data: Any) -> str:
        """Encode JSON to base64."""
        json_str = json.dumps(data)
        return EncodingUtils.to_base64(json_str)
    
    @staticmethod
    def base64_to_json(data: str) -> Any:
        """Decode JSON from base64."""
        json_str = EncodingUtils.from_base64(data)
        return json.loads(json_str)
    
    @staticmethod
    def hash_message(message: str) -> str:
        """Create message hash."""
        return hashlib.sha256(message.encode()).hexdigest()
    
    @staticmethod
    def is_valid_address(address: str, prefix: str = "deshchain") -> bool:
        """Validate address format."""
        if not address.startswith(prefix):
            return False
        if len(address) != len(prefix) + 39:
            return False
        
        address_part = address[len(prefix):]
        return address_part.isalnum() and address_part.islower()
    
    @staticmethod
    def is_valid_tx_hash(hash_str: str) -> bool:
        """Validate transaction hash."""
        return len(hash_str) == 64 and all(c in "0123456789ABCDEFabcdef" for c in hash_str)
    
    @staticmethod
    def format_amount(amount: int, decimals: int = 6) -> str:
        """Format amount with decimals."""
        return f"{amount / (10 ** decimals):.{decimals}f}"
    
    @staticmethod
    def parse_amount(amount: float, decimals: int = 6) -> int:
        """Parse amount to micro units."""
        return int(amount * (10 ** decimals))